### PR TITLE
TS-4679: Allow multicert line with action but no ssl_cert_name

### PR DESCRIPTION
**Pull request by shinrich**

### CODE REVIEW

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/450/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/344/ for details.

File: iocore/net/SSLUtils.cc [hunk 1]
```diff
@@ -2010,8 +2013,9 @@ ssl_extract_certificate(const matcher_line *line_info, ssl_user_config &sslMultC
     }
   }
   if (!sslMultCertSettings.cert) {
-    Warning("missing %s tag", SSL_CERT_TAG);
-    return false;
+    // TS-4679:  It is ok to be missing the cert.  At least if the action is set to tunnel
```
Comment (zwoop) [inline iocore/net/SSLUtils.cc]: I think you should just nuke this commented code, and maybe add a comment above about the special lenience case for the tunneling (where you check if sslMultCertSettings.cert is NULL or not).

**Comment by zwoop** [comment]
@shinrich  I think you linked this to the wrong Jira number? It should be TS-4679, right ?

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/483/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/380/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/484/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/381/ for details.

File: iocore/net/SSLUtils.cc [hunk 2]
```diff
@@ -1583,7 +1583,10 @@ SSLInitServerContext(const SSLConfigParams *params, const ssl_user_config &sslMu
     SSL_CTX_set_default_passwd_cb_userdata(ctx, &ud);
   }
 
-  if (sslMultCertSettings.cert) {
+  if (!sslMultCertSettings.cert && sslMultCertSettings.opt != SSLCertContext::OPT_TUNNEL) {
+    Warning("No ssl_cert_name specified and no tunnel action set");
+    goto fail;
+  } else if (sslMultCertSettings.cert) {
```
Comment (zwoop) [inline iocore/net/SSLUtils.cc]: I think this is "ok", it just reads a little odd. I probably would have preferred e.g.

    if (!sslMultCertSettings.cert) {
        if (sslMultCertSettings.opt != SSLCertContext::OPT_TUNNEL) {
            Warning( ...)
        }
    } else {
        SimpleTokenizer cert_tok((const ...

I know, a little nitpick, but the fear being that we repeat the conditional multiple time, having a risk of inconsistencies.

**Comment by zwoop** [comment]
:+1:

File: iocore/net/SSLUtils.cc [hunk 3]
```diff
@@ -1583,7 +1583,12 @@ SSLInitServerContext(const SSLConfigParams *params, const ssl_user_config &sslMu
     SSL_CTX_set_default_passwd_cb_userdata(ctx, &ud);
   }
 
-  if (sslMultCertSettings.cert) {
+  if (!sslMultCertSettings.cert) {
+    if (sslMultCertSettings.opt != SSLCertContext::OPT_TUNNEL) {
```
Comment (biilmann) [inline iocore/net/SSLUtils.cc]: This breaks the `TSSslServerContextCreate` plugin API method.

That method calls `SSLCreateServerContext` that delegates to `SSLInitServerContext` with an empty `sslMultCertSettings`.

This is useful for plugins that want to set the certificate/ca chain/private key themselves, for example to allow lazy loading of SNI certificates...

Comment (jpeach) [inline iocore/net/SSLUtils.cc]: @biilmann Can you please file a Jira for that?

Comment (biilmann) [inline iocore/net/SSLUtils.cc]: Done https://issues.apache.org/jira/browse/TS-4769

**Pull request merged**